test/jdk/tools/jar/JarExtractTest.java

```diff
@@ -0,0 +1,382 @@
+            printJarCommand(args);
+            int exitCode = JAR_TOOL.run(System.out, new PrintStream(err), args);
+            Assert.assertNotEquals(exitCode, 0, "jar extraction was expected to fail but didn't");
+            // verify it did indeed fail due to the right reason
+            Assert.assertTrue(err.toString(StandardCharsets.UTF_8).contains(expectedErrMsg));
+        }
+    }
+
+    /**
+     * Tests that extracting only specific files from a jar, into a specific destination directory,
+     * works as expected
+     */
+    @Test
+    public void testExtractPartialContent() throws Exception {
+        final String tmpDir = Files.createTempDirectory(Path.of("."), "8173970-").toString();
+        final String[] cmdArgs = new String[]{"-x", "-f", testJarPath.toString(), "--dir", tmpDir,
+                "f1.txt", "d1/d2/d3/f2.txt"};
+        printJarCommand(cmdArgs);
+        final int exitCode = JAR_TOOL.run(System.out, System.err, cmdArgs);
+        Assert.assertEquals(exitCode, 0, "Failed to extract " + testJarPath);
+        // make sure only the specific files were extracted
+        final Stream<Path> paths = Files.walk(Path.of(tmpDir));
+        Assert.assertEquals(paths.count(), 6, "Unexpected number of files/dirs in " + tmpDir);
+        final Path f1 = Paths.get(tmpDir, "f1.txt");
+        Assert.assertTrue(Files.isRegularFile(f1), f1.toString() + " wasn't extracted from " + testJarPath);
+        Assert.assertEquals(Files.readAllBytes(f1), FILE_CONTENT, "Unexpected content in file " + f1);
+        final Path d1 = Paths.get(tmpDir, "d1");
+        Assert.assertTrue(Files.isDirectory(d1), d1.toString() + " wasn't extracted from " + testJarPath);
+        Assert.assertEquals(Files.walk(d1, 1).count(), 2, "Unexpected number " +
+                "of files/dirs in " + d1);
+        final Path d2 = Paths.get(d1.toString(), "d2");
+        Assert.assertTrue(Files.isDirectory(d2), d2.toString() + " wasn't extracted from " + testJarPath);
+        Assert.assertEquals(Files.walk(d2, 1).count(), 2, "Unexpected number " +
+                "of files/dirs in " + d2);
+        final Path d3 = Paths.get(d2.toString(), "d3");
+        Assert.assertTrue(Files.isDirectory(d3), d3.toString() + " wasn't extracted from " + testJarPath);
+        Assert.assertEquals(Files.walk(d3, 1).count(), 2, "Unexpected number " +
+                "of files/dirs in " + d3);
+        final Path f2 = Paths.get(d3.toString(), "f2.txt");
+        Assert.assertTrue(Files.isRegularFile(f2), f2.toString() + " wasn't extracted from " + testJarPath);
+        Assert.assertEquals(Files.readAllBytes(f2), FILE_CONTENT, "Unexpected content in file " + f2);
+    }
+
+    /**
+     * Extracts the jar file using {@code jar -x -f <jarfile> -C <dest>}
+     */
+    private void testExtract(final String dest) throws Exception {
+        final String[] args = new String[]{"-x", "-f", testJarPath.toString(), "-C", dest};
```

Comment:
Perhaps add a DataProvider so you can test --extract as well?

Reply:
The latest version of the PR tests the `--extract` as well.